torch/csrc/jit/script/compiler.cpp

```diff
@@ -1820,6 +1975,49 @@ struct to_ir {
     }
   }
 
+  BoolInfo findRefinements(const TreeRef& tree) {
+    switch (tree->kind()) {
+      case TK_IS:
+      case TK_ISNOT: {
+        const auto& inputs = tree->trees();
+        if (inputs[0]->kind() == TK_VAR && inputs[1]->kind() == TK_NONE) {
+          const std::string& var_name = Var(inputs[0]).name().name();
+          Refinements true_info, false_info;
+          auto type = environment_stack->getVar(var_name, inputs[0]->range())->type();
+          if (auto opt_type = type->cast<OptionalType>()) {
+            false_info.types[var_name] = opt_type->getElementType();
```

Comment:
Why is this not a method?

Reply:
It needs access to the environment_stack

Reply:
I mean, why not `false_info.addRefinement(name, type)`?

Reply:
Semirelated - were you think that it should be an ordered_map of <string, type> and an unordered_map of <string, SourceRange> or an ordered map of <string, pair<Type, SourceRange>>? The former would have two copies of every variable... The latter is a little unwieldy but i guess that's fine